### ISSUE TITLE
Refinement for top level object failed when field has refinement

### ISSUE THREAD

**Issue by PoyangLiu** [ianstormtaylor/superstruct#979]
Using the validators below, the top level `PasswordDoNotMatch` refinement will *not* be called.
On the other hand, if we change to `newPassword: string()` (i.e. no refinement in any of the nested fields), then the `PasswordDoNotMatch` *will* be called.

```
const PasswordValidator = refine(string(), 'MinimumLength', (pw) => pw >= 8);
const ChangePasswordValidator = object({
    newPassword: PasswordValidator,
    confirmPassword: string(),
});
const RefinedChangePasswordValidator = refine(ChangePasswordValidator, 'PasswordDoNotMatch', (values) => {
    return values.newPassword == values.confirmPassword;  // This will not be called
});
```

**Comment by Gelio** [comment]
As far as I understand, this is a general problem with refinements for compound structures that they are not executed if there is **any** failure from an inner field, regardless of whether the problem is a refinement error or a validation error. The issue title is misleading - refinements for objects/arrays work fine when there are no issues inside of those objects/arrays, but are not executed when at least one refinement inside fails.

## Reproduction

In [the superstruct JSFiddle](https://jsfiddle.net/85nse1mk/)  I used the following code to reproduce the problem:

```ts
const { assert, object, number, string, refine} = window.Superstruct

const PasswordValidator = refine(string(), 'MinimumLength', (pw) => pw.length >= 8 ? true : 'required minimum length of 8');
const ChangePasswordValidator = object({
    newPassword: PasswordValidator,
    confirmPassword: string(),
});
const RefinedChangePasswordValidator = refine(ChangePasswordValidator, 'PasswordDoNotMatch', (values) => {
    return values.newPassword === values.confirmPassword ? true : "Passwords do not match";  // This will not be called
});

const data = {
  newPassword: "12345678",
  confirmPassword: '123456789'
}

try {
  assert(data, RefinedChangePasswordValidator)
  log('valid', data)
} catch (e) {
  const { message, value, type, path, branch } = e
  log('invalid', { message, value, type, path, branch })



function log(type, data) {
  document.body.className = type
  document.body.textContent = JSON.stringify(data, null, 2)
}
```

When `newPassword` was 8 characters or more, the refinement for the object run and returned the error correctly

![image](https://user-images.githubusercontent.com/889383/169279210-bc096564-d12a-484a-802b-9f1460a03427.png)

When `newPassword` was shorter than 8 characters, the only error returned was the one about the minimum length. I suppose it could return the failure for not-matching passwords too to be more complete.

## Cause of the problem

The cause of the problem is that refiner for a struct is run only when the validation of that struct (which includes validating its subfields) passes:

https://github.com/ianstormtaylor/superstruct/blob/d677b8e972d9e81d4a62f1594b55578cad55cf35/src/utils.ts#L184-L185

I suppose the solution would be to distinguish between validator failures and refinement failures. If validator was successful but there were refinement failures, we could still run the refinement for the parent object/array.

@ianstormtaylor let me know if you believe that would be a good-enough solution. This problem is what also bit me while implementing a validation that checks for uniqueness of multiple different fields in the form. It didn't run until all the other items in the form were valid. I had to do a workaround by sticking that validation inside of a struct's `validator` which runs unconditionally.

**Comment by Gelio** [comment]
Thanks for the response! Sure, I see that breaking the contract of refiners can be a major breaking change and would not be welcome. That depends on how you interpret the `value` and its subvalues being valid:

1. A `value` is valid if its `validator` and its entries' `validator`s pass
2. A `value` is valid if its and its entries' `validator`s and `refiners`s pass

Not sure which one is the more correct definition of _valid_. The former definition is looser but allows collecting errors from multiple refiners for compound values.

If we don't want to change the default behavior, maybe we should introduce a new refiner type that is looser? That doesn't sound great too, but at least it is not a breaking change.

Anyway, it's not a priority for me, since I get by with the workaround of adding my custom refiner into the `validator`. We could make that solution more standard by adding some APIs to define these _loose refiners_ easier.

**Comment by ianstormtaylor** [comment]
@Gelio yeah, I see what you mean. Thinking about it more, I think it would be okay to make this change, and like you say have refiners be given a value that has been validated but not refined. If someone really didn't want this, they can always define their refinements as validations instead to get really strict. Otherwise what's the point of refinements being a separate concept really.

If you wanted to pull request that change I'd be down to merge it!

**Comment by Gelio** [comment]
Great! Sure, I'll try to prepare something over the weekend.

Just to be clear, I will disambiguate the errors so that we have specific `validator` and `refiner` errors and then run the `refiner` if there were no `validator` errors on the `value` itself or its entries.

**Comment by ianstormtaylor** [comment]
@Gelio that sounds great! Actually I believe there is already an `error.refinement` field which might give us exactly what we need to make that check in the `run` function.

**Comment by Gelio** [comment]
I took a stab at this problem and delivered https://github.com/ianstormtaylor/superstruct/pull/1068
